File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque


def solution(bridge_length, weight, truck_weights):
    bridge_q = deque([0] * bridge_length)
    truck_q = deque(truck_weights)
    current_weight = 0
    time_cnt = 0

    while bridge_q:
        current_weight -= bridge_q.popleft()
        time_cnt += 1

        if truck_q:
            if weight >= current_weight + truck_q[0]:
                truck = truck_q.popleft()
                current_weight += truck
                bridge_q.append(truck)

            else:
                bridge_q.append(0)

    return time_cnt

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Deque<Integer> dq = new ArrayDeque<>();
        for(int i = 0; i < bridge_length; i++){
            dq.offer(0);
        }
        int N = truck_weights.length;
        int curW = 0;
        int tIdx = 0;
        int time = 0;
        while(true){
            curW -= dq.poll();
            time++;
            if(tIdx < N && curW + truck_weights[tIdx] <= weight){
                int truckW = truck_weights[tIdx];
                curW += truckW;
                dq.offer(truckW);
                tIdx++;
            }else {
                dq.offer(0);
            }
            if(curW == 0 && tIdx == N){
                break;
            }
            
            
        }

        return time;
    }
}